# Projeto 3 - Ciência dos Dados

#  Projeto UFC

Nome: Victor Vergara Arcoverde de Albuquerque Cavalcanti

Nome: Edgard Neto

Nome: Gabriel Yamashita

Nome: Henrique Mualem Marti



  ___
## Objetivo:

### O objetivo desse projeto é fazer um machine learning para poder prever qual lutador ganhará uma luta do UFC baseado no seu histórico. Para isso serão usados os dados de todas as lutas do UFC (mais de 5 mil) a fim de descobrir quais os fatores dos lutadores que impactam mais no resultado das lutas. 

[Database utilizado](https://www.kaggle.com/rajeevw/ufcdata#data.csv)

 ___
## Preparando o ambiente no jupyter:

### Imports:

In [2]:
import math
import os.path
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import json
import random
import statsmodels.api as sm
import seaborn as sns

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import MultinomialNB

from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

### Trabalhando com os Excels:

In [3]:
data = pd.read_excel("data.xlsx")
preproc = pd.read_excel("preprocessed_data.xlsx")
fighter = pd.read_excel("raw_fighter_details.xlsx")
total = pd.read_excel("raw_total_fight_data.xlsx")

data.head(2)

R_fighter      B_fighter          Referee  date  \
0          Henry Cejudo  Marlon Moraes     Marc Goddard  2019   
1  Valentina Shevchenko    Jessica Eye  Robert Madrigal  2019   

                 location Winner  title_bout       weight_class  no_of_rounds  \
0  Chicago, Illinois, USA    Red        True       Bantamweight             5   
1  Chicago, Illinois, USA    Red        True  Women's Flyweight             5   

   B_current_lose_streak  ...    R_win_by_KO/TKO  R_win_by_Submission  \
0                      0  ...                  2                    0   
1                      0  ...                  0                    2   

   R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
0                             0       8  Orthodox        162.56       162.56   
1                             0       5  Southpaw        165.10       167.64   

   R_Weight_lbs  B_age  R_age  
0         135.0   31.0   32.0  
1         125.0   32.0   31.0  

[2 rows x 145 columns]

In [4]:
data.weight_class = data.weight_class.astype('category')
data.Winner = data.Winner.astype('category')

In [5]:
data.Winner.value_counts()

Red     3470
Blue    1591
Draw      83
Name: Winner, dtype: int64

In [6]:
data_heavy = data.loc[(data.weight_class=='Heavyweight'),:]
bool_to_number = {False: 0, True: 1}
string_to_number = {'Blue': 0, 'Red': 1, 'Draw': 2}
data_heavy['title_bout'] = data_heavy['title_bout'].map(bool_to_number)
data_heavy['Winner'] = data_heavy['Winner'].map(string_to_number)

C:\Users\gabri\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\Users\gabri\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


#### Blue = 0
#### Red = 1

In [7]:
data_heavy.head(5)

R_fighter          B_fighter          Referee  date  \
4       Tai Tuivasa      Blagoy Ivanov  Dan Miragliotta  2019   
54      Walt Harris     Serghei Spivac      Jerin Valel  2019   
60    Arjan Bhullar         Juan Adams     Yves Lavigne  2019   
63       Greg Hardy  Dmitrii Smoliakov   Keith Peterson  2019   
69  Andrei Arlovski      Augusto Sakai  Dan Miragliotta  2019   

                   location  Winner  title_bout weight_class  no_of_rounds  \
4    Chicago, Illinois, USA       0           0  Heavyweight             3   
54  Ottawa, Ontario, Canada       1           0  Heavyweight             3   
60  Ottawa, Ontario, Canada       1           0  Heavyweight             3   
63    Sunrise, Florida, USA       1           0  Heavyweight             3   
69    Sunrise, Florida, USA       0           0  Heavyweight             3   

    B_current_lose_streak  ...    R_win_by_KO/TKO  R_win_by_Submission  \
4                       0  ...                  2                    0   
54                      0  ...                  4                    0   
60                      0  ...                  0                    0   
63                      2  ...                  0                    0   
69                      0  ...                  9                    2   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
54                             0       4  Southpaw        195.58       195.58   
60                             0       2  Orthodox        185.42       190.50   
63                             0       0  Orthodox        195.58       203.20   
69                             0      16  Orthodox        190.50       195.58   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
54         250.0   24.0   35.0  
60         245.0   27.0   32.0  
63         265.0   36.0   30.0  
69         240.0   27.0   40.0  

[5 rows x 145 columns]

In [7]:
data_heavy_util = data_heavy.drop(['Referee','date','location'], axis=1)
#dados que não se relacionam com os lutadores ou seus resultado

In [8]:
data_heavy_util.head()

R_fighter          B_fighter  Winner  title_bout weight_class  \
4       Tai Tuivasa      Blagoy Ivanov       0           0  Heavyweight   
54      Walt Harris     Serghei Spivac       1           0  Heavyweight   
60    Arjan Bhullar         Juan Adams       1           0  Heavyweight   
63       Greg Hardy  Dmitrii Smoliakov       1           0  Heavyweight   
69  Andrei Arlovski      Augusto Sakai       0           0  Heavyweight   

    no_of_rounds  B_current_lose_streak  B_current_win_streak  B_draw  \
4              3                      0                     1       0   
54             3                      0                     0       0   
60             3                      0                     1       0   
63             3                      2                     0       0   
69             3                      0                     1       0   

    B_avg_BODY_att  ...    R_win_by_KO/TKO  R_win_by_Submission  \
4             17.0  ...                  2                    0   
54             NaN  ...                  4                    0   
60             8.0  ...                  0                    0   
63             5.0  ...                  0                    0   
69            35.0  ...                  9                    2   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
54                             0       4  Southpaw        195.58       195.58   
60                             0       2  Orthodox        185.42       190.50   
63                             0       0  Orthodox        195.58       203.20   
69                             0      16  Orthodox        190.50       195.58   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
54         250.0   24.0   35.0  
60         245.0   27.0   32.0  
63         265.0   36.0   30.0  
69         240.0   27.0   40.0  

[5 rows x 142 columns]

In [9]:
data_heavy_util.dropna(inplace=True)
data_heavy_util.head()

R_fighter          B_fighter  Winner  title_bout weight_class  \
4        Tai Tuivasa      Blagoy Ivanov       0           0  Heavyweight   
60     Arjan Bhullar         Juan Adams       1           0  Heavyweight   
63        Greg Hardy  Dmitrii Smoliakov       1           0  Heavyweight   
69   Andrei Arlovski      Augusto Sakai       0           0  Heavyweight   
75  Alistair Overeem    Aleksei Oleinik       1           0  Heavyweight   

    no_of_rounds  B_current_lose_streak  B_current_win_streak  B_draw  \
4              3                      0                     1       0   
60             3                      0                     1       0   
63             3                      2                     0       0   
69             3                      0                     1       0   
75             5                      0                     2       0   

    B_avg_BODY_att  ...    R_win_by_KO/TKO  R_win_by_Submission  \
4            17.00  ...                  2                    0   
60            8.00  ...                  0                    0   
63            5.00  ...                  0                    0   
69           35.00  ...                  9                    2   
75            4.25  ...                  6                    0   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
60                             0       2  Orthodox        185.42       190.50   
63                             0       0  Orthodox        195.58       203.20   
69                             0      16  Orthodox        190.50       195.58   
75                             0       9  Orthodox        193.04       203.20   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
60         245.0   27.0   32.0  
63         265.0   36.0   30.0  
69         240.0   27.0   40.0  
75         265.0   41.0   38.0  

[5 rows x 142 columns]

#### Tirando os dados categóricos:

In [10]:
categoricas = [
    'R_fighter', 
    'B_fighter', 
    'weight_class', 
    'R_Stance', 
    'B_Stance', 
]

data_heavy_cat = data_heavy_util[categoricas].astype('category')
data_heavy_num = data_heavy_util.drop(categoricas, axis=1).astype('float')

In [11]:
X = data_heavy_num.drop('Winner', axis=1)
y = data_heavy_num['Winner']

### Separando os dados em testes e treinamento

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [13]:
model = RandomForestClassifier(n_estimators=10000)
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10000, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [14]:
#Fatores que tem o maior peso na decisão da vitória
j=1
lista_j = list()
for i, f in sorted(list(zip(model.feature_importances_, X_train.columns)), reverse=True):
    print( j,'°:',f'{f}: {i}')
    a=str(j)+'°'
    lista_j.append(a)
    j+=1


1 °: R_avg_opp_GROUND_att: 0.014189377968060265
2 °: B_avg_HEAD_att: 0.013721741236613795
3 °: R_avg_HEAD_landed: 0.013671778917814936
4 °: R_avg_opp_GROUND_landed: 0.013582513216252486
5 °: B_avg_TOTAL_STR_att: 0.013206834774490315
6 °: B_avg_BODY_att: 0.012670645025397921
7 °: B_avg_HEAD_landed: 0.012659041513257138
8 °: B_avg_BODY_landed: 0.01245525660893241
9 °: B_avg_SIG_STR_landed: 0.012180765884913666
10 °: B_avg_SIG_STR_att: 0.012175325843527277
11 °: R_losses: 0.01194116687000994
12 °: R_age: 0.0119095364767234
13 °: R_avg_opp_KD: 0.011858188192011121
14 °: R_avg_HEAD_att: 0.011464917151220003
15 °: R_avg_SIG_STR_landed: 0.011166433862636046
16 °: R_total_rounds_fought: 0.010984720158479221
17 °: B_avg_opp_TOTAL_STR_landed: 0.010805375366328102
18 °: R_avg_opp_HEAD_landed: 0.01079560784912558
19 °: B_avg_TOTAL_STR_landed: 0.010668416059572455
20 °: R_avg_GROUND_landed: 0.01065597033149818
21 °: R_avg_SIG_STR_att: 0.010209465857228012
22 °: B_avg_GROUND_landed: 0.01018466836539

#### Criando um dataframe com os dados acima:

In [15]:
data={'Fator':X_train.columns ,'Correlação':model.feature_importances_,}
Fator_por_corr=pd.DataFrame(data)
Fator_por_corr=Fator_por_corr.sort_values(by='Correlação', ascending=False)
Fator_por_corr['Grau de Importância']=lista_j
Fator_por_corr = Fator_por_corr.set_index('Grau de Importância')
Fator_por_corr.head()

Fator  Correlação
Grau de Importância                                     
1°                      R_avg_opp_GROUND_att    0.014189
2°                            B_avg_HEAD_att    0.013722
3°                         R_avg_HEAD_landed    0.013672
4°                   R_avg_opp_GROUND_landed    0.013583
5°                       B_avg_TOTAL_STR_att    0.013207

___
## Verificando a performance


In [16]:
X_train.columns

Index(['title_bout', 'no_of_rounds', 'B_current_lose_streak',
       'B_current_win_streak', 'B_draw', 'B_avg_BODY_att', 'B_avg_BODY_landed',
       'B_avg_CLINCH_att', 'B_avg_CLINCH_landed', 'B_avg_DISTANCE_att',
       ...
       'R_win_by_Decision_Unanimous', 'R_win_by_KO/TKO', 'R_win_by_Submission',
       'R_win_by_TKO_Doctor_Stoppage', 'R_wins', 'R_Height_cms', 'R_Reach_cms',
       'R_Weight_lbs', 'B_age', 'R_age'],
      dtype='object', length=136)

In [17]:
model.feature_importances_

array([0.00050141, 0.00098601, 0.00246992, 0.00287439, 0.        ,
       0.01267065, 0.01245526, 0.00935895, 0.00929696, 0.00932483,
       0.00740604, 0.0101518 , 0.01018467, 0.01372174, 0.01265904,
       0.00517554, 0.0081571 , 0.00797083, 0.00601562, 0.00551591,
       0.01217533, 0.01218077, 0.00813521, 0.00427077, 0.00938571,
       0.00691786, 0.00696899, 0.01320683, 0.01066842, 0.00361877,
       0.00373991, 0.00968804, 0.00969772, 0.00778071, 0.00765114,
       0.00743412, 0.00687237, 0.00986881, 0.00962792, 0.00826573,
       0.00860674, 0.00475234, 0.00923278, 0.00817709, 0.00754315,
       0.00460561, 0.00760215, 0.0072777 , 0.00924938, 0.00459431,
       0.00697802, 0.00642157, 0.00865915, 0.0085013 , 0.01080538,
       0.00656783, 0.0077386 , 0.00178878, 0.00039555, 0.00103158,
       0.00238398, 0.00447065, 0.00151515, 0.00039427, 0.0040145 ,
       0.00531592, 0.00965348, 0.00911479, 0.00384247, 0.00539589,
       0.        , 0.00871214, 0.00824541, 0.00882592, 0.00849

In [18]:
y_pred = model.predict(X_test)

In [19]:
print(accuracy_score(y_test, y_pred))

0.6666666666666666


In [20]:
y_pred

array([1., 1., 1., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 1., 1., 0., 1.,
       1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1.])

In [21]:
y_test.value_counts(True)

1.0    0.636364
0.0    0.363636
Name: Winner, dtype: float64

In [22]:
data_heavy_num.shape

(264, 137)

### Conclusão desse teste:

Tendo um acurácia de 0.6667 não é um bom resultado, visto que o modelo praticamente sempre tem como resultado o vermelho como vencedor, e como a probabilidade do lutador vermelho ganhar é de 0.6212112 ele praticamente só acerta os que o vencedor é o vermelho e erra os que o azul é o vencedor.

Assim é possível concluir que é necessário desconsiderar algumas variáveis para tentar melhorar a acurácia.

___
## Montando a Regressão Logística


## !Ainda é necessário tirar os fatores que foram julgados ter um baixo grau de importância pelo teste anterior!

In [23]:
def preparo(X,Y):
    X_cp = sm.add_constant(X)
    model = sm.OLS(Y,X_cp,missing='drop')
    results = model.fit()
    return results

In [24]:
Y = data_heavy_num["Winner"]
data_heavy_sem_Winner=data_heavy_num.drop('Winner',axis=1) 
X=data_heavy_sem_Winner
#np.asarray(X)

In [25]:
result = preparo(X,Y)
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Winner   R-squared:                       0.515
Model:                            OLS   Adj. R-squared:                  0.120
Method:                 Least Squares   F-statistic:                     1.305
Date:                Thu, 14 Nov 2019   Prob (F-statistic):             0.0635
Time:                        14:04:32   Log-Likelihood:                -100.66
No. Observations:                 264   AIC:                             439.3
Df Residuals:                     145   BIC:                             864.9
Df Model:                         118                                         
Covariance Type:            nonrobust                                         
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const                            4.6535      2.927      1.590      0.114      -1.131      10.438
title_bout                      -0.0643      0.192     -0.334      0.739      -0.445       0.316
no_of_rounds                     0.0958      0.062      1.533      0.127      -0.028       0.219
B_current_lose_streak           -0.2204      0.073     -3.002      0.003      -0.366      -0.075
B_current_win_streak            -0.0361      0.037     -0.983      0.327      -0.109       0.037
B_draw                        1.562e-13   4.62e-14      3.377      0.001    6.48e-14    2.48e-13
B_avg_BODY_att                  -0.0701      0.040     -1.752      0.082      -0.149       0.009
B_avg_BODY_landed                0.0788      0.049      1.623      0.107      -0.017       0.175
B_avg_CLINCH_att                 0.0259      0.025      1.025      0.307      -0.024       0.076
B_avg_CLINCH_landed             -0.0234      0.035     -0.669      0.504      -0.092       0.046
B_avg_DISTANCE_att              -0.0151      0.014     -1.059      0.291      -0.043       0.013
B_avg_DISTANCE_landed            0.0141      0.019      0.727      0.469      -0.024       0.053
B_avg_GROUND_att                -0.0245      0.023     -1.059      0.291      -0.070       0.021
B_avg_GROUND_landed              0.0242      0.032      0.764      0.446      -0.038       0.087
B_avg_HEAD_att                  -0.0283      0.022     -1.283      0.201      -0.072       0.015
B_avg_HEAD_landed                0.0352      0.028      1.266      0.208      -0.020       0.090
B_avg_KD                         0.3686      0.193      1.906      0.059      -0.014       0.751
B_avg_LEG_att                    0.0846      0.059      1.436      0.153      -0.032       0.201
B_avg_LEG_landed                -0.0990      0.071     -1.386      0.168      -0.240       0.042
B_avg_PASS                       0.1028      0.075      1.375      0.171      -0.045       0.251
B_avg_REV                        0.7377      0.295      2.497      0.014       0.154       1.322
B_avg_SIG_STR_att               -0.0137      0.023     -0.606      0.546      -0.059       0.031
B_avg_SIG_STR_landed             0.0150      0.027      0.551      0.583      -0.039       0.069
B_avg_SIG_STR_pct               -1.5196      0.575     -2.644      0.009      -2.656      -0.384
B_avg_SUB_ATT                   -0.1195      0.137     -0.871      0.385      -0.391       0.152
B_avg_TD_att                    -0.0131      0.048     -0.274      0.784      -0.107       0.081
B_avg_TD_landed                 -0.0729      0.104     -0.701      0.484      -0.278       0.133
B_avg_TD_pct                     0.3372      0.275      1.228      0.222      -0.206       0.880
B_avg_TOTAL_STR_att              0.0405      0.026      1.568      0.119    

In [26]:
'''data_heavy_blue_winners = data_heavy.loc[(data_heavy.Winner==0),:]
data_heavy_blue_winners = data_heavy_blue_winners.drop(['Winner'], axis=1)
data_heavy_red_winners = data_heavy.loc[(data_heavy.Winner==1),:]
data_heavy_red_winners = data_heavy_red_winners.drop(['Winner'], axis=1)'''

"data_heavy_blue_winners = data_heavy.loc[(data_heavy.Winner==0),:]\ndata_heavy_blue_winners = data_heavy_blue_winners.drop(['Winner'], axis=1)\ndata_heavy_red_winners = data_heavy.loc[(data_heavy.Winner==1),:]\ndata_heavy_red_winners = data_heavy_red_winners.drop(['Winner'], axis=1)"

In [27]:
'''blue_winners_correlation = data_heavy_blue_winners.corr()
blue_winners_correlation.head()'''

'blue_winners_correlation = data_heavy_blue_winners.corr()\nblue_winners_correlation.head()'

In [28]:
'''red_winners_correlation = data_heavy_red_winners.corr()
red_winners_correlation.head()'''

'red_winners_correlation = data_heavy_red_winners.corr()\nred_winners_correlation.head()'

   ___
## Referências

[Como usar a biblioteca Scikit-lear](https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression)

[Como funciona o Random Forest](https://towardsdatascience.com/understanding-random-forest-58381e0602d2)

[Referencia do Random Forest Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

[Como funciona regressão linear](https://www.saedsayad.com/logistic_regression.htm)

[Como usar a Regressão Linear](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)

